src/reanimated2/animations.js

```diff
@@ -354,55 +354,83 @@ export function withDecay(userConfig, callback) {
     'worklet';
     const config = {
       deceleration: 0.998,
+      velocityFactor: 1,
     };
     if (userConfig) {
       Object.keys(userConfig).forEach((key) => (config[key] = userConfig[key]));
     }
 
-    const VELOCITY_EPS = 5;
+    const VELOCITY_EPS = 1;
+    const SLOPE_FACTOR = 0.1;
 
     function decay(animation, now) {
-      const { lastTimestamp, initialVelocity, current, velocity } = animation;
+      const {
+        lastTimestamp,
+        startTimestamp,
+        initialVelocity,
+        current,
+        velocity,
+      } = animation;
 
       const deltaTime = Math.min(now - lastTimestamp, 64);
-      animation.lastTimestamp = now;
-
-      const kv = Math.pow(config.deceleration, deltaTime);
-      const kx = (config.deceleration * (1 - kv)) / (1 - config.deceleration);
-
-      const v0 = velocity / 1000;
-      const v = v0 * kv * 1000;
-      const x = current + v0 * kx;
-
-      animation.current = x;
+      const v =
+        velocity *
+        Math.exp(
+          -(1 - config.deceleration) * (now - startTimestamp) * SLOPE_FACTOR
+        );
+      animation.current =
+        current + (v / (1000 * config.velocityFactor)) * deltaTime;
       animation.velocity = v;
+      animation.lastTimestamp = now;
 
-      let toValueIsReached = null;
-
-      if (Array.isArray(config.clamp)) {
+      if (config.clamp) {
         if (initialVelocity < 0 && animation.current <= config.clamp[0]) {
-          toValueIsReached = config.clamp[0];
+          animation.current = config.clamp[0];
+          return true;
         } else if (
           initialVelocity > 0 &&
           animation.current >= config.clamp[1]
         ) {
-          toValueIsReached = config.clamp[1];
+          animation.current = config.clamp[1];
+          return true;
         }
       }
 
-      if (Math.abs(v) < VELOCITY_EPS || toValueIsReached !== null) {
-        if (toValueIsReached !== null) {
-          animation.current = toValueIsReached;
-        }
-
+      if (Math.abs(v) < VELOCITY_EPS) {
         return true;
       }
     }
 
+    function validConfig() {
```

Comment:
TBH I'd rewrite this like so:

```js
if (!Array.isArray(config.clamp)) {
	throw new Error("not an array");
}
if (config.clamp.length !== 2) {
	throw new Error("invalid number of arguments")
}
if (config.velocityFactor <= 0) {
	throw new Error("invalid velocity");
}
```
Since we actually want to developer to fix their code instead of showing errors in the console.
Inverting conditions makes it a little more readable IMHO.

Reply:
Also, this should be named `validateConfig()` instead.

Reply:
Yes, I agree that exceptions can be more readable but there is a problem. This is `worklet` and if I throw an exception in `worklet` function the user receives RedBox with `"Javascript worklet error in file reanimated2/core.js"` instead of my error message. IMO this information is useless and confusing, so that's why I use `console.error` instead of an exception.

We should resolve the problem with exceptions but this is the case for the next PR.

Reply:
Ah, I see. Makes sense to leave `console.error` in that case.